lang/java/idl/src/main/java/org/apache/avro/util/IdlUtils.java

```diff
@@ -0,0 +1,541 @@
+    final String newlinePlusIndent = NEWLINE + indent + " * ";
+    while (matcher.find()) {
+      if (!foundMatch) {
+        buffer.append(newlinePlusIndent);
+        foundMatch = true;
+      }
+      matcher.appendReplacement(buffer, newlinePlusIndent);
+    }
+    if (foundMatch) {
+      matcher.appendTail(buffer);
+      buffer.append(NEWLINE).append(indent).append(" */").append(NEWLINE);
+    } else {
+      buffer.append(doc).append(" */").append(NEWLINE);
+    }
+    return buffer.toString();
+  }
+
+  private static void writeFieldSchema(Schema schema, Writer writer, Set<String> alreadyDeclared, Set<Schema> toDeclare,
+      String recordNameSpace) throws IOException {
+    Schema.Type type = schema.getType();
+    if (type == Schema.Type.RECORD || type == Schema.Type.ENUM || type == Schema.Type.FIXED) {
+      if (Objects.equals(recordNameSpace, schema.getNamespace())) {
+        writer.append(schema.getName());
+      } else {
+        writer.append(schema.getFullName());
+      }
+      if (!alreadyDeclared.contains(schema.getFullName())) {
+        toDeclare.add(schema);
+      }
+    } else if (type == Schema.Type.ARRAY) {
+      writeJsonProperties(schema, writer, null);
+      writer.append("array<");
+      writeFieldSchema(schema.getElementType(), writer, alreadyDeclared, toDeclare, recordNameSpace);
+      writer.append('>');
+    } else if (type == Schema.Type.MAP) {
+      writeJsonProperties(schema, writer, null);
+      writer.append("map<");
+      writeFieldSchema(schema.getValueType(), writer, alreadyDeclared, toDeclare, recordNameSpace);
+      writer.append('>');
+    } else if (type == Schema.Type.UNION) {
+      // Note: unions cannot have properties
+      Schema schemaForNullableSyntax = getNullableUnionType(schema);
+      if (schemaForNullableSyntax != null) {
+        writeFieldSchema(schemaForNullableSyntax, writer, alreadyDeclared, toDeclare, recordNameSpace);
+        writer.append('?');
+      } else {
+        writer.append("union{");
+        List<Schema> types = schema.getTypes();
+        Iterator<Schema> iterator = types.iterator();
+        if (iterator.hasNext()) {
+          writeFieldSchema(iterator.next(), writer, alreadyDeclared, toDeclare, recordNameSpace);
+          while (iterator.hasNext()) {
+            writer.append(", ");
+            writeFieldSchema(iterator.next(), writer, alreadyDeclared, toDeclare, recordNameSpace);
+          }
+        } else {
+          throw new AvroRuntimeException("Union schemas must have member types " + schema);
+        }
+        writer.append('}');
+      }
+    } else {
+      Set<String> propertiesToSkip = new HashSet<>();
+      String typeName;
+      if (schema.getLogicalType() == null) {
+        typeName = schema.getName();
+      } else {
+        String logicalName = schema.getLogicalType().getName();
+        switch (logicalName) {
+        case "date":
+        case "time-millis":
+        case "timestamp-millis":
```

Comment:
I think recently added `nanos` is missing here

Reply:
For a later PR, I would prefer using constants here. This would be easy if we make the ones in `LogicalTypes` public: https://github.com/apache/avro/blob/ab7ac6ce6bc5cfc5e04791415ae3bb77fc29fd80/lang/java/avro/src/main/java/org/apache/avro/LogicalTypes.java#L200

Reply:
Agreed. There's more refactoring to do anyway.

Reply:
> I think recently added `nanos` is missing here

Correct: the IDL format does not support those natively, yet. Nor the `-micros` versions. This is something that can be added in another PR.